File: python/lib/rssi_reader.py
from __future__ import annotations

import subprocess


def get_rssi() -> tuple[int | None, int | None, float | None, float | None]:
    """Run netsh wlan show interfaces and parse signal, RSSI, and data rates.

    Handles both Turkish and English locale output.
    Returns (signal_percent, rssi_dbm, receive_rate_mbps, transmit_rate_mbps).
    Never raises; returns (None, None, None, None) on any failure.
    """
    try:
        result = subprocess.run(
            ["netsh", "wlan", "show", "interfaces"],
            capture_output=True, text=True, encoding="utf-8", errors="ignore", timeout=5,
        )
        if result.returncode == 0:
            sig, dbm, rx, tx = None, None, None, None
            for line in result.stdout.split("\n"):
                ll = line.lower()
                if ("signal" in ll or "sinyal" in ll) and "%" in line:
                    try:
                        sig = int(line.split(":")[1].replace("%", "").strip())
                    except Exception:
                        pass
                elif "rssi" in ll and "bssid" not in ll:
                    try:
                        dbm = int(line.split(":")[1].strip())
                    except Exception:
                        pass
                elif "receive rate" in ll or "al\u0131m h\u0131z\u0131" in ll or ("al" in ll and "mbps" in ll):
                    try:
                        rx = float(line.split(":")[1].replace("Mbps", "").strip())
                    except Exception:
                        pass
                elif "transmit rate" in ll or "iletim h\u0131z\u0131" in ll or ("ilet" in ll and "mbps" in ll):
                    try:
                        tx = float(line.split(":")[1].replace("Mbps", "").strip())
                    except Exception:
                        pass
            return sig, dbm, rx, tx
        return None, None, None, None
    except Exception:
        return None, None, None, None


def can_read_rssi() -> bool:
    """Check whether netsh wlan can be invoked without permission errors."""
    try:
        r = subprocess.run(
            ["netsh", "wlan", "show", "interfaces"],
            capture_output=True, text=True, encoding="utf-8", errors="ignore", timeout=5,
        )
        out = (r.stdout + r.stderr).lower()
        return "location permission" not in out and "access is denied" not in out
    except Exception:
        return False

File: WifiMotionDotNet/WifiMotion/Core/RssiReader.cs
using System;
using System.Diagnostics;
using System.Globalization;
using System.Text;

namespace WifiMotion.Core;

/// <summary>Tek bir RSSI okumasinin sonucu.</summary>
public readonly record struct RssiReading(int? Signal, int? Dbm, double? Rx, double? Tx);

/// <summary>
/// <c>netsh wlan show interfaces</c> ciktisini ayristirarak sinyal, RSSI ve veri
/// hizlarini okur. Python <c>lib/rssi_reader.py</c> modulunun portu.
/// Hem Turkce hem Ingilizce yerel ayar ciktisini destekler.
/// </summary>
public static class RssiReader
{
    private const int TimeoutMs = 5000;

    /// <summary>
    /// <c>netsh wlan show interfaces</c> calistirir; (sinyal%, RSSI dBm, alim Mbps,
    /// iletim Mbps) dondurur. Asla istisna firlatmaz; basarisizlikta hepsi null'dur.
    /// </summary>
    public static RssiReading GetRssi()
    {
        var (ok, stdout, _) = RunNetsh();
        if (!ok)
            return new RssiReading(null, null, null, null);

        int? sig = null, dbm = null;
        double? rx = null, tx = null;

        foreach (string line in stdout.Split('\n'))
        {
            string ll = line.ToLowerInvariant();
            int colon = line.IndexOf(':');
            string after = colon >= 0 && colon + 1 < line.Length ? line[(colon + 1)..] : "";

            if ((ll.Contains("signal") || ll.Contains("sinyal")) && line.Contains('%'))
            {
                if (TryExtractInt(after, out int v))
                    sig = v;
            }
            else if (ll.Contains("rssi") && !ll.Contains("bssid"))
            {
                if (TryExtractInt(after, out int v))
                    dbm = v;
            }
            else if (ll.Contains("receive rate") || ll.Contains("alım hızı") ||
                     (ll.Contains("al") && ll.Contains("mbps")))
            {
                if (TryExtractDouble(after, out double v))
                    rx = v;
            }
            else if (ll.Contains("transmit rate") || ll.Contains("iletim hızı") ||
                     (ll.Contains("ilet") && ll.Contains("mbps")))
            {
                if (TryExtractDouble(after, out double v))
                    tx = v;
            }
        }

        return new RssiReading(sig, dbm, rx, tx);
    }

    /// <summary>
    /// <c>netsh wlan</c> izin hatasi olmadan calistirilabilir mi kontrol eder.
    /// Python <c>can_read_rssi</c> karsiligi.
    /// </summary>
    public static bool CanReadRssi()
    {
        var (ok, stdout, stderr) = RunNetsh();
        if (!ok)
            return false;
        string outAll = (stdout + stderr).ToLowerInvariant();
        return !outAll.Contains("location permission") && !outAll.Contains("access is denied");
    }

    private static (bool Ok, string Stdout, string Stderr) RunNetsh()
    {
        try
        {
            var psi = new ProcessStartInfo
            {
                FileName = "netsh",
                Arguments = "wlan show interfaces",
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true,
                StandardOutputEncoding = Encoding.UTF8,
                StandardErrorEncoding = Encoding.UTF8,
            };
            using var proc = Process.Start(psi);
            if (proc is null)
                return (false, "", "");

            string stdout = proc.StandardOutput.ReadToEnd();
            string stderr = proc.StandardError.ReadToEnd();
            if (!proc.WaitForExit(TimeoutMs))
            {
                try { proc.Kill(true); } catch { /* yoksay */ }
                return (false, "", "");
            }
            return (proc.ExitCode == 0, stdout, stderr);
        }
        catch (Exception)
        {
            return (false, "", "");
        }
    }

    /// <summary>Bir metin parcasindan ilk tam sayiyi (negatif olabilir) cikarir.</summary>
    private static bool TryExtractInt(string s, out int value)
    {
        var sb = new StringBuilder();
        foreach (char c in s)
        {
            if (char.IsDigit(c) || (c == '-' && sb.Length == 0))
                sb.Append(c);
            else if (sb.Length > 0)
                break;
        }
        return int.TryParse(sb.ToString(), NumberStyles.Integer, CultureInfo.InvariantCulture, out value);
    }

    /// <summary>Bir metin parcasindan ilk ondalik sayiyi cikarir ('.' veya ',' destekli).</summary>
    private static bool TryExtractDouble(string s, out double value)
    {
        var sb = new StringBuilder();
        foreach (char c in s)
        {
            if (char.IsDigit(c) || (c == '-' && sb.Length == 0))
                sb.Append(c);
            else if ((c == '.' || c == ',') && sb.Length > 0)
                sb.Append('.');
            else if (sb.Length > 0)
                break;
        }
        return double.TryParse(sb.ToString(), NumberStyles.Float, CultureInfo.InvariantCulture, out value);
    }
}
